### PR TITLE
Task/run 33 modal start run

### PR DESCRIPTION
**Pull request by sam-kakengi**
Using AI to generate a summary of what I've done! Take a look.

Note: Need to eventually work on managing how refreshing will work with the paused duration, but it works with the dummy timer. Buttons are still not fully adjusting to them though.

## Key Changes

1. **Improved Run Start Logic**
   - Added error handling and validation for the API response when starting a run.
   - Implemented fetching of route details to ensure the correct route name is displayed.

2. **Enhanced Run End Functionality**
   - Introduced a new `endRun` function that updates the run details via a PATCH request.
   - Added cleanup of local storage items related to runs upon ending a run.

3. **Paused Run Management**
   - Implemented state management for paused runs, including duration tracking.

4. **Modal Management**
   - Added state management for end run and pre-end run modals.

5. **Context API Expansion**
   - Extended the ActiveRunContext to include new functions and state variables.

## Technical Details
- Utilized `localStorage` for persisting run start times and managing paused run states.
- Implemented error handling and logging for API calls and data fetching operations.
- Expanded the context provider to expose new functionalities to child components.

### CODE REVIEW

**Comment by Dimwiddle** [comment]
@sam-kakengi  Noticed that the active run dashboard is not matching the design. Can you make the adjustments with the colours etc?

Figma design [here](https://www.figma.com/design/QeuZ0v64aPp5Hd7SjU8QO1/TrackRunner?node-id=7634-2317&node-type=frame&t=vxTnnRamM6i4AzBz-0)

**Actual**
image

**Expected**

image

**Comment by sam-kakengi** [comment]
Will do 👍🏽

**Comment by Dimwiddle** [comment]
@sam-kakengi  Just some feedback on the pre-modal. These changes would be good just for consistency with the other modals.

- Red comments are text changes
- White comments are style edits 

image

File: frontend/src/components/core-components/context/ActiveRunV2.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                },
+                body: JSON.stringify({
+                    route: routeID,
+                    notes: notes
+                })
+            })
+      
+            if (!response.ok) {
+                throw new Error(`HTTP error! status: ${response.status}`)
+            }
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/context/ActiveRunV2.js:121]: You can save some lines here by handling this request in the `apiClient.js`. Feel free to create a shortcut method if `patchData()` is not enough

File: frontend/src/components/core-components/dashboard/EndRunModal.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+          
+          
+          const matchingRoute = retrievedRoutes.find(route => route.name === recentRunData.route)
+          setRunData({
+            routeID: matchingRoute ? matchingRoute.id : '',
+            routeName: recentRunData.route,
+            notes: recentRunData.notes || '',
+          })
+        } catch (error) {
+          console.error('Error fetching data:', error)
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/EndRunModal.js:49]: Just keeping a generic error message is fine here. No need to show `error`

File: frontend/src/components/core-components/dashboard/PreEndRunModal.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+  return (
+    <ThemeProvider theme={modalTheme}>
+      <Dialog open={open} onClose={onCancel} fullWidth maxWidth="sm">
+        <DialogContent sx={{ backgroundColor: modalTheme.palette.secondary.main }}>
+          <Paper elevation={0} sx={{ padding: { xs: 2, sm: 3, md: 4 }, margin: 'auto', backgroundColor: modalTheme.palette.secondary.main }}>
+            <Grid container spacing={isMobile ? 2 : 3}>
+              <Grid item xs={12}>
+                <Typography variant={isMobile ? 'h6' : 'h5'} sx={{ color: modalTheme.palette.primary.main }} gutterBottom>
+                  End Run Confirmation
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/PreEndRunModal.js:30]: Change to _Complete Run_

File: frontend/src/components/core-components/dashboard/PreEndRunModal.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                </Typography>
+              </Grid>
+              <Grid item xs={12}>
+                <Typography variant="body1" sx={{ color: 'black', fontWeight: 'bold' }}>
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/PreEndRunModal.js:34]: stick to the palette for styling. Possibly bluegrey[50] could do the job here. 

Bold comes across a bit too strong. A slightly heavier weight than _normal_ might look better.

File: frontend/src/components/core-components/dashboard/PreEndRunModal.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                  Are you sure you want to end your current run?
+                </Typography>
+              </Grid>
+              <Grid item xs={12}>
+                <DialogActions sx={{ padding: 0, justifyContent: 'flex-start', gap: 2 }}>
+                  <Button onClick={handleConfirm} variant="contained" color="primary" size={isMobile ? 'small' : 'medium'} fullWidth={isMobile}>
+                    Confirm End Run
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/PreEndRunModal.js:41]: Change to: _End Run_

File: frontend/src/components/core-components/dashboard/startRunComponents/activeRunTileV2.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        }
+    }, [pausedRun.isPaused, pausedRun.pausedDuration])
+
+    const formatDuration = (duration) => {
+        const hours = Math.floor(duration / 3600)
+        const minutes = Math.floor((duration % 3600) / 60)
+        const seconds = duration % 60
+      
+        return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`
+    }
```
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/startRunComponents/activeRunTileV2.js:49]: Methods like these are better stored in `utilities/` for easier re-use.

File: frontend/src/components/core-components/dashboard/startRunComponents/startRunModalV2.js
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/startRunComponents/startRunModalV2.js:1]: Any reason for V2 naming?

File: frontend/src/components/core-components/dashboard/startRunComponents/activeRunTileV2.js
Comment (Dimwiddle) [inline frontend/src/components/core-components/dashboard/startRunComponents/activeRunTileV2.js:1]: Any reason for V2 naming?

**Review by Dimwiddle — CHANGES_REQUESTED** [review]
Overall it's looking great! Really nice work. Just some changes to look at before merging this in.

**Pull request merged**